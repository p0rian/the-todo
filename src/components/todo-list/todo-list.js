import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  addTodo,
  selectAllTodos,
  selectActiveTodos,
  selectCompletedTodos,
} from "store/slices/todoSlice";
import { TodoItem } from "components/todo-item/todo-item";
import { TodoFilter } from "components/todo-filter/todo-filter";
import { Paper, Input, Button } from "@mui/material";

import css from "./todo-list.module.css";

export const TodoList = () => {
  const [newItem, setNewItem] = useState("");
  const [filter, setFilter] = useState("all");

  const dispatch = useDispatch();

  const todos = useSelector((state) => {
    switch (filter) {
      case "active":
        return selectActiveTodos(state);
      case "completed":
        return selectCompletedTodos(state);
      default:
        return selectAllTodos(state);
    }
  });

  const handleAddClick = () => {
    if (newItem.trim() !== "") {
      setNewItem("");
      dispatch(addTodo(newItem));
    }
  };

  const handleFilterChange = (_, newFilter) => {
    setFilter(newFilter);
  };

  return (
    <div className={css.appContainer}>
      <Paper className={css.listContainer}>
        <div className={css.inputWrapper}>
          <Input
            className={css.inputStyle}
            value={newItem}
            type="text"
            onChange={(e) => setNewItem(e.target.value)}
          ></Input>
          <Button onClick={handleAddClick}>Add Item</Button>
        </div>
        <div className={css.todoList}>
          <ul className={css.listStyle}>
            {todos.map((item) => (
              <TodoItem key={item.id} item={item} />
            ))}
          </ul>
          <TodoFilter filter={filter} handleFilterChange={handleFilterChange} />
        </div>
      </Paper>
    </div>
  );
};
