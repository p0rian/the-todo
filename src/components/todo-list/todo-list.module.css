.appContainer {
  background: #f2f2f2;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
}

.listContainer {
  width: 400px;
  padding: 20px;
}

.inputWrapper {
  display: flex;
  justify-content: space-between;
  margin-bottom: 20px;
}

.inputStyle {
  width: 70%;
}

.addButton {
  width: 25%;
}

.todoList {
  list-style: none;
  margin: 0;
  padding: 0;
}

.filterStyle {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.counterContainer {
  display: flex;
  flex-direction: column;
  margin-right: 10px;
}

.counterStyle {
  margin-bottom: 5px;
}

.listStyle {
  padding-left: 0;
  list-style: none;
}
