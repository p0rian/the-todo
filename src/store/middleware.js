import { createListenerMiddleware, isAnyOf } from "@reduxjs/toolkit";
import { addTodo, updateTodo, removeTodo } from "./slices/todoSlice";

export const listenerMiddleware = createListenerMiddleware();
listenerMiddleware.startListening({
  matcher: isAnyOf(addTodo, updateTodo, removeTodo),
  effect: (_, listenerApi) => {
    localStorage.setItem("todoData", JSON.stringify(listenerApi.getState()));
  },
});
