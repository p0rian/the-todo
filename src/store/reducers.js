import { todoReducer } from "./slices/todoSlice";

export const reducers = {
  todoData: todoReducer,
};
