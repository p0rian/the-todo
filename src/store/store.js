import { configureStore } from "@reduxjs/toolkit";
import { reducers } from "./reducers";
import { listenerMiddleware } from "./middleware";

const todoState = JSON.parse(localStorage.getItem("todoData") || "null");
const preloadedState = todoState ? { todoData: todoState.todoData } : {};

export const store = configureStore({
  preloadedState,
  reducer: { ...reducers },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(listenerMiddleware.middleware),
});
