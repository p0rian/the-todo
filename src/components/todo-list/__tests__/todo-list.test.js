import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { store } from "store/store";
import { TodoList } from "../todo-list";

describe("TodoList component", () => {
  it("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <TodoList />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
