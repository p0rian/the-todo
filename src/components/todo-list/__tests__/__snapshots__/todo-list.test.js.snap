// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component renders correctly 1`] = `
<div>
  <div
    class="appContainer"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 listContainer css-1ps6pg7-MuiPaper-root"
    >
      <div
        class="inputWrapper"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary inputStyle css-q0jhri-MuiInputBase-root-MuiInput-root"
        >
          <input
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            type="text"
            value=""
          />
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Add Item
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="todoList"
      >
        <ul
          class="listStyle"
        />
      </div>
      <div
        class="filterStyle"
      >
        <div
          class="counterContainer"
        >
          <div
            class="counterStyle"
          >
            Total items:
            0
          </div>
          <div
            class="counterStyle"
          >
            Active items left:
            0
          </div>
        </div>
        <div
          class="MuiBox-root css-1og1qar"
        >
          <div
            aria-label="todo filter"
            class="MuiToggleButtonGroup-root css-1gjgmky-MuiToggleButtonGroup-root"
            role="group"
          >
            <button
              aria-label="show all todos"
              aria-pressed="true"
              class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-ueukts-MuiButtonBase-root-MuiToggleButton-root"
              tabindex="0"
              type="button"
              value="all"
            >
              All
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="show active todos"
              aria-pressed="false"
              class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-ueukts-MuiButtonBase-root-MuiToggleButton-root"
              tabindex="0"
              type="button"
              value="active"
            >
              Active
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="show completed todos"
              aria-pressed="false"
              class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-ueukts-MuiButtonBase-root-MuiToggleButton-root"
              tabindex="0"
              type="button"
              value="completed"
            >
              Completed
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
